File: 0155-min-stack/0155-min-stack.js

class Node{
    constructor(val){
        this.val = val;
        this.next = null;
    }
}
var MinStack = function() {
    // const newNode = new Node(null);
    this.stackTop = null;
    this.length = 0;
};

/** 
 * @param {number} val
 * @return {void}
 */
MinStack.prototype.push = function(val) {
    const newNode = new Node(val);
    if(this.stackTop === null){
        this.stackTop = newNode;
    }
    else{
        newNode.next = this.stackTop;
        this.stackTop = newNode;
    }
    this.length++;
    return null;
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function() {
    if(this.stackTop){
        let currNode = this.stackTop;
        this.stackTop = currNode.next;
        currNode.next = null;
        this.length--;
    }
    return null;
};

/**
 * @return {number}
 */
MinStack.prototype.top = function() {
    if(this.stackTop === null) return null;
    return this.stackTop.val;
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function() {
    if(this.stackTop){
        let currNode = this.stackTop;
        let minVal = currNode.val;
        
        while(currNode){
            currNode = currNode.next;
            if(currNode){
                let newMinVal = currNode.val;
                if(newMinVal < minVal){
                    minVal = newMinVal;
                }
            }
        }
        return minVal;
    }
    return null;
};

/** 
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(val)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */

File: 0155-min-stack/0155-min-stack.py
class MinStack:

    def __init__(self):
        self.min_stack = []
        self.size = 0

    def push(self, val: int) -> None:
        self.min_stack.append(val)
        self.size += 1        

    def pop(self) -> None:
        if(self.size == 0):
            return None
        else:
            self.min_stack.pop()
            self.size -= 1        

    def top(self) -> int:
        if(self.size == 0):
            return None
        else:
            return self.min_stack[self.size - 1]        

    def getMin(self) -> int:
        if(self.size == 0):
            return None
        else:
            return min(self.min_stack)    

# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()